src/main/java/cms/gov/madie/measure/services/MeasureService.java

```diff
@@ -0,0 +1,50 @@
+package cms.gov.madie.measure.services;
+
+import cms.gov.madie.measure.exceptions.ResourceNotFoundException;
+import cms.gov.madie.measure.models.Group;
+import cms.gov.madie.measure.models.Measure;
+import cms.gov.madie.measure.repositories.MeasureRepository;
+import org.bson.types.ObjectId;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class MeasureService {
+  private final MeasureRepository measureRepository;
+
+  public MeasureService(MeasureRepository measureRepository) {
+    this.measureRepository = measureRepository;
+  }
+
+  public Group createOrUpdateGroup(Group group, String measureId, String username) {
+    Measure measure = measureRepository.findById(measureId).orElse(null);
+    if (measure == null) {
+      throw new ResourceNotFoundException("Measure", measureId);
+    }
+    // no group present, this is the first group
+    if (CollectionUtils.isEmpty(measure.getGroups())) {
+      group.setId(ObjectId.get().toString());
```

Comment:
Just curious, is this needed for add?

Reply:
I think so. spring mongo data doesn't create an Id for Embedded documents. Assigning id manually helps to identify the subdocument by id during updates.